File: asst6/1.java
import java.util.Scanner;
class Gen extends Thread
{	
	// find divisors
	Object lock = new Object();
	int num;
	int sum = 0;
	Gen(int n)
	{
		this.num = n;
	}
	public void run()
	{
		synchronized(this)
		{
			try
			{
				for(int i=1;i<num;i++)
				{
					if(num%i==0)
						sum+=i;
				}
				wait();
				Thread.sleep(100);
				notify();
			}
			catch(InterruptedException e)
			{
				e.printStackTrace();
			}
			
		}
	}
	public int getSum()
	{
		synchronized(this)
		{
			try
			{
				notify();
				wait();
			}
			catch(InterruptedException e)
			{
				e.printStackTrace();
			}
		}
		return sum;
		
	}
}
class AmicableNumber
{
	public static void main(String[] args)
	{
		Scanner sc = new Scanner(System.in);
		int a = sc.nextInt();
		int b = sc.nextInt();
		Gen csa = new Gen(a);
		Gen csb = new Gen(b);
		csa.start();
		csb.start();
		if(csa.getSum() == b && csb.getSum() == a)
		{
			System.out.println("Amicable numbers.");
		}
		else
		{
			System.out.println("Not amicable numbers.");
		}
	}
}

File: asst6/1.py
import threading
import time

first=0
second=0
s1=0
s2=0
def Generate1(lock):
    lock.acquire()
    global first,s1
    print("Thread name:",threading.current_thread().name,end="  ")
    
    for i in range (1,first):
        if(first%i ==0):
            s1+=i
    lock.release()

    
    
def Generate2(lock):
    lock.acquire()
    global second,s2
    print("Thread name:",threading.current_thread().name,end="  ")
    for i in range (1,second):
        if(second%i ==0):
            s2+=i
    lock.release()

if __name__ == "__main__":
    lock=threading.Lock()
	first=int(input("Enter first number:"))
    second=int(input("Enter second number:"))
    t1=threading.Thread(target=Generate1, args=(lock,),name="CSA")
    t2=threading.Thread(target=Generate2, args=(lock,),name="CSB")

    t1.start()
    # time.sleep(2)
    t2.start()
    t1.join()
    t2.join()
    print(first,second,s1,s2)
    if(first==s2 and second==s1):
        print("Numbers are amicable")
    else:
        print("Numbers are not amicable")
        

    
    
